will/xmpp_plugins.py

```diff
@@ -0,0 +1,289 @@
+import base64
+import ssl
+import logging
+
+from sleekxmpp.util import bytes, sasl
+from sleekxmpp.util.stringprep_profiles import StringPrepError
+from sleekxmpp.stanza import StreamFeatures
+from sleekxmpp.xmlstream import StanzaBase, RestartStream, register_stanza_plugin
+from sleekxmpp.plugins import BasePlugin
+from sleekxmpp.xmlstream.matcher import MatchXPath
+from sleekxmpp.xmlstream.handler import Callback
+from sleekxmpp.features.feature_mechanisms import stanza
+import settings
+
+log = logging.getLogger(__name__)
+
+
+class CustomAuth(stanza.Auth):
+    """
+    """
+
+    name = 'auth'
+    # namespace = 'urn:ietf:params:xml:ns:xmpp-sasl'
+    namespace = 'http://hipchat.com'
+    interfaces = set(('mechanism', 'value'))
+    plugin_attrib = name
+
+    #: Some SASL mechs require sending values as is,
+    #: without converting base64.
+    plain_mechs = set(['X-MESSENGER-OAUTH2'])
+
+    def setup(self, xml):
+        StanzaBase.setup(self, xml)
+        self.xml.tag = self.tag_name()
+        self.xml.set("oauth2_token", "true")
+
+    def get_value(self):
+        if not self['mechanism'] in self.plain_mechs:
+            return base64.b64decode(bytes(self.xml.text))
+        else:
+            return self.xml.text
+
+    def set_value(self, values):
+        if not self['mechanism'] in self.plain_mechs:
+            values = "0%s0%s0will" % (
```

Comment:
VICTORY!! I've got an oauth2_token in the response! 

Now I just have to figure out how to pull it out via xmpp, but that should be *much* simpler, and something that's tackleable with a fresh brain in the morning.

@dougkeen thanks much for your help, and the quick responses from everyone there at hipchat.  My kudos and gratitude.